code/controllers/Processes/scheduler.dm

```diff
@@ -4,50 +4,89 @@
 * Scheduler *
 ************/
 /datum/controller/process/scheduler
-	var/list/scheduled_tasks
-	var/tick_completed = TRUE
-	var/list/queued_tasks
+	var/list/datum/scheduled_task/tasks = list()
+	var/datum/scheduled_task/head
 
 /datum/controller/process/scheduler/setup()
 	name = "scheduler"
-	schedule_interval = 2 SECONDS
-	scheduled_tasks = list()
+	schedule_interval = 2
 	scheduler = src
 
 /datum/controller/process/scheduler/doWork()
-	if (tick_completed)
-		queued_tasks = scheduled_tasks.Copy()
-		tick_completed = FALSE
-
-	while (queued_tasks.len)
-		var/datum/scheduled_task/task = queued_tasks[queued_tasks.len]
-		queued_tasks.len--
-
-		if (QDELETED(task))
-			scheduled_tasks -= task
+	while (head && !PSCHED_CHECK_TICK)
+		if (head.destroyed)
+			head = head.next
+			tasks -= head
+			head.kill()
+			continue
+		if (head.trigger_time >= world.time)
+			return	// Nothing after this will be ready to fire.
+
+		// This one's ready to fire, process it.
+		var/datum/scheduled_task/task = head
+		head = task.next
+		task.pre_process()
+		task.process()
+		task.post_process()
+		if (task.destroyed)	// post_process probably destroyed it.
+			tasks -= task
+			task.kill()
+
+/datum/controller/process/scheduler/proc/queue(datum/scheduled_task/task)
+	if (!task || !task.trigger_time)
+		warning("scheduler: Invalid task queued! Ignoring.")
+		return
+	// Reset this in-case we're doing a rebuild.
+	task.next = null
+	if (!head && !tasks.len)
+		head = task
+		tasks += task
+		return
+
+	if (!head)	// Head's missing but we still have tasks, rebuild.
+		tasks += task
+		rebuild_queue()
+		return
+
+	var/datum/scheduled_task/curr = head
+	while (curr.next && curr.trigger_time < task.trigger_time)
+		curr = curr.next
+
+	if (!curr.next)
+		// We're at the end of the queue, just append.
+		curr.next = task
+		tasks += task
+		return
+	
+	// Inserting midway into the list.
+	var/old_next = curr.next
+	curr.next = task
+	task.next = old_next
+	tasks += task
+
+// Rebuilds the queue linked-list, removing invalid or destroyed entries.
+/datum/controller/process/scheduler/proc/rebuild_queue()
+	log_debug("Rebuilding queue.")
```

Comment:
`log_debug()` messages originating from a consistent source should carry a short label.

Examples:
`API:`
`BOREALIS:`
`MIRRORS:`